stackslib/src/clarity_vm/database/mod.rs

```diff
@@ -316,16 +533,92 @@ where
     })
 }
 
-fn get_miner_column<F, R>(
+fn get_stacks_header_column<F, R>(
+    conn: &DBConn,
+    id_bhh: &StacksBlockId,
+    column_name: &str,
+    loader: F,
+) -> Option<R>
+where
+    F: Fn(&Row) -> R,
+{
+    match get_stacks_header_column_from_table(conn, id_bhh, column_name, &loader, false) {
+        Some(x) => Some(x),
+        None => get_stacks_header_column_from_table(conn, id_bhh, column_name, &loader, true),
+    }
+}
+
+fn get_first_block_in_tenure(
     conn: &DBConn,
     id_bhh: &StacksBlockId,
+    epoch_opt: Option<&StacksEpochId>,
+) -> TenureBlockId {
+    let consensus_hash = match epoch_opt {
+        Some(epoch) => {
+            if !epoch.uses_nakamoto_blocks() {
+                return id_bhh.clone().into();
+            } else {
+                get_stacks_header_column_from_table(
+                    conn,
+                    id_bhh,
+                    "consensus_hash",
+                    &|r| ConsensusHash::from_row(r).expect("FATAL: malformed consensus_hash"),
+                    true,
+                )
+            }
+        }
+        None => {
+            if let Some(ch) = get_stacks_header_column_from_table(
+                conn,
+                id_bhh,
+                "consensus_hash",
+                &|r| ConsensusHash::from_row(r).expect("FATAL: malformed consensus_hash"),
+                false,
+            ) {
+                return id_bhh.clone().into();
+            } else {
+                get_stacks_header_column_from_table(
+                    conn,
+                    id_bhh,
+                    "consensus_hash",
+                    &|r| ConsensusHash::from_row(r).expect("FATAL: malformed consensus_hash"),
+                    true,
+                )
+            }
+        }
+    };
+    let ch = consensus_hash
+        .expect("Unexpected SQL failure querying block header table for 'consensus_hash'");
+    let args: &[&dyn ToSql] = &[&ch];
+    conn.query_row(
```

Comment:
Just a heads-up: this is going to break with my work on MARF'ing the chainstate.  With malleablized signatures, it's possible for multiple, different `StacksBlockId`s to be returned for a given tenure's start block.  This code will need to be revised when this change is merged.  Can you open an issue for that?

Reply:
#4900 👍